CONTRIBUTING.md

```diff
@@ -24,15 +24,149 @@ Workspaces][workspaces], with the main crates being:
+- `run-fail` - running the Rune should fail (due to a missing capability,
+  crash, etc.)
+
+Let's go through the process of adding an integration test which evaluates
+the `microspeech` Rune and makes sure it can detect the word, "up".
+
+The first thing to do is create a folder for it:
+
+```console
+$ mkdir integration-tests/run-pass/microspeech-up
+$ cd integration-tests/run-pass/microspeech-up
+```
+
+Next we need to add our Runefile. Normally you'd write it from scratch but
+because we are already using it as an example, we'll just symlink it into the
+`microspeech-up/` directory.
+
+```console
+$ ln -s ../../../examples/microspeech/Runefile.yml ./Runefile.yml
+```
+
+This Rune also requires a `model.tflite` so we'll symlink that too.
+
+```console
+$ ln -s ../../../examples/microspeech/model.tflite ./model.tflite
+```
+
+If this was just testing `rune build` (i.e. a `compile-pass` or `compile-fail`
+test) we could stop here. However, because we want to actually run the Rune we
+need to supply it with any capabilities it will need.
+
+The way this works is pretty simple. Say you would normally run the Rune with
+`rune run ./microspeech-up.rune --sound up.wav` then just add an `up.wav` file
+to the test directory.
+
+```console
+$ ln -s ../../../examples/microspeech/data/up/84d1e469_nohash_0.wav ./up.wav
+```
+
+Capabilities are determined based on their extension, so a `*.png` will be
+passed in using `--image`, `*.wav` files are `--audio`, and so on.
+
+By default, the only things that get checked are the exit code from `rune`.
+
+|             | **pass**            | **fail**           |
+| ----------- | ------------------- | ------------------ |
+| **compile** | `rune build` passes | `rune build` fails |
+| **rune**    | `rune run` passes   | `rune run` fails   |
+
+However, you can add additional checks to make sure particular strings are
+printed to the screen. This is really useful when trying to improve error
+messages and the end-user's experience, but for now let's just make sure
+the Rune correctly outputs "up".
+
+The first thing to do is run the Rune manually.
+
+```console
+$ rune build ./Runefile.yml
+$ rune run microspeech-up.rune --sound up.wav
+[2021-05-27T15:36:17.117Z INFO  rune::run] Running rune: microspeech-up.rune
+[2021-05-27T15:36:17.153Z DEBUG rune_wasmer_runtime] Loading image
+[2021-05-27T15:36:17.153Z DEBUG rune_wasmer_runtime] Instantiating the WebAssembly module
+[2021-05-27T15:36:17.153Z DEBUG rune_wasmer_runtime] Loaded the Rune
+[2021-05-27T15:36:17.154Z DEBUG rune_wasmer_runtime] Running the rune
+[2021-05-27T15:36:17.163Z INFO  runicos_base::image] Serial: {"type_name":"&str","channel":2,"elements":["up"],"dimensions":[1]}
+```
+
+We just care about that last log message so let's copy it to a new file. The
+file can be called whatever you want as long as it has the `stderr` extension.
+
+```console
+$ cat expected.stderr
+Serial: {"type_name":"&str","channel":2,"elements":["up"],"dimensions":[1]}
```

Comment:
How does this work for multi channel outputs?

Reply:
It'd be the same as normal. Multiple outputs are just returned as a JSON array.

```console
$ cat Runefile.yml
image: runicos/base

pipeline:
  byte_input:
    capability: RAW
    outputs:
      - type: u8
        dimensions: [16]
  integer_input:
    capability: RAW
    outputs:
      - type: i32
        dimensions: [4]

  output:
    out: SERIAL
    inputs:
    - byte_input
    - integer_input

$ python3
>>> import sys
>>> import struct
>>> with open("input.bin", "wb") as f:
           f.write(struct.pack("IIII", 0, 1, 2, 3))
>>> exit()

$ xxd input.bin
00000000: 0000 0000 0100 0000 0200 0000 0300 0000  ................

$ cargo run --package rune -- run multiple-outputs.rune --raw input.bin
    Finished dev [unoptimized + debuginfo] target(s) in 0.09s
     Running `/home/michael/Documents/hotg-ai/rune/target/debug/rune run multiple-outputs.rune --raw input.bin`
[2021-05-27T17:10:37.882Z INFO  rune::run] Running rune: multiple-outputs.rune
[2021-05-27T17:10:38.227Z DEBUG rune_wasmer_runtime] Loading image
[2021-05-27T17:10:38.228Z DEBUG rune_wasmer_runtime] Instantiating the WebAssembly module
[2021-05-27T17:10:38.230Z DEBUG rune_wasmer_runtime] Loaded the Rune
[2021-05-27T17:10:38.230Z DEBUG rune_wasmer_runtime] Running the rune
[2021-05-27T17:10:38.230Z INFO  runicos_base::image] Serial: [
    {"type_name":"u8","channel":2,"elements":[0,0,0,0,1,0,0,0,2,0,0,0,3,0,0,0],"dimensions":[16]},
    {"type_name":"i32","channel":2,"elements":[0,1,2,3],"dimensions":[4]}
]